File: de.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define PRINT_DOTS
#define PRINT_NEWBEST
//#define PRINT_MOVEMENTS
#define PRINT_FINAL_ANSWER

typedef double (*objective_fn)(void* object);
typedef void (*guess_fn)(double* vector, size_t N);

void unibox_guess(double* v, size_t N) {
    int i;
    for(i=0;i<N;i++) {
        v[i] = drand48();
    }
}

void print_vec(double* v, size_t N) {
    int i;
    printf("[ ");
    for(i=0;i<N;i++) {
        printf("%-10.5g ",v[i]);
    }
    printf("]");
}

double differential_evolution(double* result, size_t N, objective_fn f, double F, double CR, int NP, guess_fn guess, int iters) {
    if(!(F >= 0 && F <= 2)) {
        F = 1.0;
    }
    if(!(CR >= 0 && CR <= 1)) {
        CR = 0.8;
    }
    if(NP < 4) {
        NP = 4;
    }
    if(guess == NULL) {
        guess = unibox_guess;
    }
    double* xs = malloc(sizeof(double)*N*NP);
    int i;
    for(i=0;i<NP;i++) {
        guess(xs+(i*N),N);
    }
    double bestf = INFINITY;
    int bestx = 0;
    double fs[NP];
    for(i=0;i<NP;i++) {
        fs[i] = f(xs+i*N);
    }
    do {
        int x;
        for(x=0;x<NP;x++) {
            double *xp = xs+x*N;
            int a,b,c;
            do { a = lrand48() % NP; } while(a == x);
            do { b = lrand48() % NP; } while(b == x || b == a);
            do { c = lrand48() % NP; } while(c == x || c == a || c == b);
            int R = lrand48() % N;
            double y[N];
            for(i=0; i<N; i++) {
               double r = drand48();
               if(r < CR || i == R) {
                   y[i] = xs[a*N+i] + F*(xs[b*N+i] - xs[c*N+i]);
               } else {
                   y[i] = xp[i];
               }
            }
            double fy = f(y);
            double fx = fs[x];
            if(fy < fx) {
#ifdef PRINT_MOVEMENTS
                printf("Moving #%d from ",x);
                print_vec(xp,N);
                printf(" (f: %e) to ", fx);
                print_vec(y,N);
                printf(" (f: %e)\n", fy);
#endif
                memcpy(xp,y,sizeof(double)*N);
                fs[x] = fy;
                if(fy < bestf) {
                    bestf = fy;
                    bestx = x;
#ifdef PRINT_NEWBEST
                    printf("\nNew best f(x): %e ",bestf);
                    print_vec(xp,N);
                    printf("\n");
#endif
                }
            }
        }
    } while(--iters);
    memcpy(result,xs+bestx*N,sizeof(double)*N);
    return bestf;
}

double tgt[] = {1.0, -2.0, 3.0, -4.0, 5.0, -6.0, 7.0, -8.0};
size_t N = sizeof(tgt)/sizeof(double);

int total_f_evals = 0;

double test_f(void* o) {
    total_f_evals++;
#ifdef PRINT_DOTS
    printf(".");
#endif
    double* v = (double*)o;
    int i;
    double acc = 0.0;
    for(i=0; i<N; i++) {
        double diff = v[i] - tgt[i];
        acc += diff*diff;
    }
    return acc;
}

int main(int argc, char** argv) {
    int iters = 1e2;
    double F=1.0, CR=0.0, NP=40;
    if(argc>1) {
        iters = atoi(argv[1]);
        if(argc>2) {
            NP = atoi(argv[2]);
            if(argc>3) {
                CR = atof(argv[3]);
                if(argc>4) {
                    F = atof(argv[4]);
                }}}}

    srand48(time(NULL));
    double result[N];   
    double bestf = differential_evolution(result, N, test_f, F, CR, NP, NULL, iters);
#ifdef PRINT_FINAL_ANSWER
    printf("\nFINAL X  f(X): %e ",bestf);
    print_vec(result,N);
    printf("\nwith %d function evaluations\n", total_f_evals);
#endif
    return 0;
}

File: de.lua
local ffi = require("ffi")

PRINT_DOTS = false
PRINT_NEWBEST = false
PRINT_MOVEMENTS = false
PRINT_FINAL_ANSWER = true

local function unibox_guess(v, N)
    for i = 0, N-1 do
        v[i] = math.random()
    end
end

local function print_vec(v, N)
    io.write '[ '
    for i = 0, N-1 do
        io.write(string.format("%-10.5g ",v[i]))
    end
    io.write ']'
end

local doubles = ffi.typeof("double[?]")

local function differential_evolution(result, N, f, F, CR, NP, guess, iters)
    if (not (F and F >= 0 and F <= 2)) then
        F = 1.0
    end
    if (not (CR and CR >= 0 and CR <= 1)) then
        CR = 0.8
    end
    if (not (NP and NP >= 4)) then
        NP = 4
    end
    if (not guess) then
        guess = unibox_guess
    end
    
    local xs = ffi.new(doubles, N*NP)

    for x = 0, NP-1 do
        guess(xs+(x*N),N)
    end
    
    local y = ffi.new(doubles, N)

    local bestf = tonumber("infinity")
    local bestx = 0
    local fs = ffi.new(doubles, NP)
    for x = 0, NP-1 do
        fs[x] = f(xs+x*N)
    end
    repeat
        for x = 0, NP-1 do
            local xp = xs+x*N
            local a,b,c
            repeat a = math.random(0,NP-1) until (a ~= x)
            repeat b = math.random(0,NP-1) until (b ~= x and b ~= a)
            repeat c = math.random(0,NP-1) until (c ~= x and c ~= a and c ~= b)
            local R = math.random(0,N-1)
            for i = 0, N-1 do
                r = math.random()
                if (r < CR or i == R) then
                    y[i] = xs[a*N+i] + F*(xs[b*N+i] - xs[c*N+i])
                else
                    y[i] = xp[i]
                end
            end
            local fy = f(y)
            local fx = fs[x]
            if(fy < fx) then
                if(PRINT_MOVEMENTS) then
                    io.write(string.format("Moving #%d from ",x))
                    print_vec(xp,N)
                    io.write(string.format(" (f: %e) to ",fx))
                    print_vec(y, N)
                    io.write(string.format(" (f: %e)\n",fy))
                end
                ffi.copy(xp,y,ffi.sizeof(doubles,N))
                fs[x] = fy
                if(fy < bestf) then
                    bestf = fy
                    bestx = x
                    if(PRINT_NEWBEST) then
                        io.write(string.format("\nNew best f(x): %e ",bestf))
                        print_vec(xp, N)
                        io.write("\n")
                    end
                end
            end
        end
        iters = iters - 1
    until (iters == 0)
    ffi.copy(result,xs+bestx*N,ffi.sizeof(doubles,N))   
    return bestf
end

local tgt = {[0]=1.0, -2.0, 3.0, -4.0, 5.0, -6.0, 7.0, -8.0, 9.0, -10.0, 11.0, -12.0, 13.0, -14.0, 15.0, -16.0, 17.0, -18.0, 19.0, -20.0, 21.0, -22.0, 23.0, -24.0, 25.0, -26.0, 27.0, -28.0, 29.0, -30.0, 31.0, -32.0}
local N = #tgt

local total_f_evals = 0

local function test_f(v)
    total_f_evals = total_f_evals + 1
    if(PRINT_DOTS) then
        io.write(".")
    end
    local acc = 0.0
    for i = 0, N-1 do
        local diff = v[i] - tgt[i]
        acc = acc + diff*diff
    end
    return acc
end

local iters = 1100
local runs = 50
local F, CR, NP = 1.0, 0.0, 40
if arg[1] then
    iters = tonumber(arg[1])
    if arg[2] then
        runs = tonumber(arg[2])
        if arg[3] then
            NP = tonumber(arg[3])
            if arg[4] then
                CR = tonumber(arg[4])
                if arg[5] then
                    F = tonumber(arg[5])
                end end end end end

math.randomseed(os.time())
local result = ffi.new(doubles,N)
repeat
    local bestf = differential_evolution(result,N,test_f,F,CR,NP,nil,iters)
    if(PRINT_FINAL_ANSWER) then
        io.write(string.format("\nFINAL X  f(X): %e ",bestf))
        print_vec(result,N)
        io.write(string.format("\nwith %d total function evaluations\n", total_f_evals))
    end
    runs = runs - 1
until runs == 0
